src/plugins/initiatives/nodeEntry.js

```diff
@@ -0,0 +1,45 @@
+// @flow
+
+import {type URL} from "../../core/references";
+import {type NodeWeight} from "../../core/weights";
+import {type TimestampMs, type TimestampISO} from "../../util/timestamp";
+
+/**
+ * Represents an "inline contribution" node. They're called entries and named
+ * by type: contribution entry, reference entry, dependency entry.
+ * The generalization of this is a node entry.
+ */
+
+// Which field the NodeEntry appeared in.
+export type NodeEntryField = "DEPENDENCY" | "REFERENCE" | "CONTRIBUTION";
+
+// A normalized NodeEntry type, guaranteed to have all relevant fields set.
+export type NodeEntry = {|
+  // A single line description of what this contribution represents.
+  +title: string,
+  // A unique key to use as a suffix to the NodeAddress.
+  // The InitiativeFile and NodeEntryField will be prefixes, so it must be
+  // unique within one field's array of entries.
+  +key: string,
+  // URLs pointing to the contributors. Similar to the initiatives will use
+  // reference detection to create edges.
+  +contributors: $ReadOnlyArray<URL>,
+  // Timestamp for this node.
+  +timestampMs: TimestampMs,
+  // NodeWeight for this node.
+  +weight: NodeWeight | null,
```

Comment:
Weights have their own defaulting mechanism, the prefix multiplications.

So to minimize graph bloat and computations, instead of adding a noop `1x` as default for each node, we'll fall back on the type's weight. That's what `null` means in this case, not adding a weight for this node.

I did intentionally write: "guaranteed to have all relevant fields set" (e.g. set to null is not unset) rather than guaranteed to be non-null. Perhaps the comment should be changed to avoid the confusion.

Most of the `NodeEntryJson` optionals are there for flexibility and ease of user input. For example, a missing `contributors` key, will default to an empty array. It's easier when both of these are considered a noop. The business logic will expect there to always be an array, so the empty array is the "correct" internal noop.

The nullable weight is not a user convenience, it's to support a business logic case. It gives you control over whether a custom weight (ex: 20x), a noop weight (1x) or no weight (null) should be added to the graph. `NodeWeight | null` is the minimal representation of these approaches.

Reply:
Note: this is the same rationale as we're using for the `Initiative` type and came up in a discussion there. https://github.com/sourcecred/sourcecred/pull/1674#pullrequestreview-366148374